tests/AcmeClientTest.php

```diff
@@ -0,0 +1,147 @@
+use AcmePhp\Ssl\Generator\KeyPairGenerator;
+use AcmePhp\Ssl\Parser\KeyParser;
+use AcmePhp\Ssl\Signer\DataSigner;
+use GuzzleHttp\Client;
+use Symfony\Component\Process\PhpExecutableFinder;
+use Symfony\Component\Process\Process;
+
+class AcmeClientTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @var AcmeClientInterface
+     */
+    private $client;
+
+    public function setUp()
+    {
+        $secureHttpClient = new SecureHttpClient(
+            (new KeyPairGenerator())->generateKeyPair(),
+            new Client(),
+            new Base64SafeEncoder(),
+            new KeyParser(),
+            new DataSigner(),
+            new ServerErrorHandler()
+        );
+
+        $this->client = new AcmeClient($secureHttpClient, 'http://127.0.0.1:4000/directory');
+    }
+
+    /**
+     * @expectedException \AcmePhp\Core\Exception\Server\MalformedServerException
+     */
+    public function testDoubleRegisterAccountFail()
+    {
+        $this->client->registerAccount();
+        $this->client->registerAccount();
+    }
+
+    /**
+     * @expectedException \AcmePhp\Core\Exception\Server\MalformedServerException
+     */
+    public function testInvalidAgreement()
+    {
+        $this->client->registerAccount('http://invalid.com');
+        $this->client->requestAuthorization('example.org');
+    }
+
+    public function testFullProcess()
+    {
+        /*
+         * Register account
+         */
+        $data = $this->client->registerAccount('http://boulder:4000/terms/v1');
+
+        $this->assertInternalType('array', $data);
+        $this->assertArrayHasKey('id', $data);
+        $this->assertArrayHasKey('key', $data);
+        $this->assertArrayHasKey('initialIp', $data);
+        $this->assertArrayHasKey('createdAt', $data);
+
+        /*
+         * Ask for domain challenge
+         */
+        $challenge = $this->client->requestAuthorization('acmephp.com');
+
+        $this->assertInstanceOf('AcmePhp\\Core\\Protocol\\AuthorizationChallenge', $challenge);
+        $this->assertEquals('acmephp.com', $challenge->getDomain());
+        $this->assertContains('http://127.0.0.1:4000/acme/challenge', $challenge->getUrl());
+        $this->assertContains('http://127.0.0.1:4000/acme/authz', $challenge->getLocation());
+
+        /*
+         * Challenge check
+         */
+        $process = $this->createServerProcess($challenge);
+        $process->start();
+
+        $this->assertTrue($process->isRunning());
+
+        $check = $this->client->challengeAuthorization($challenge);
+
+        $this->assertEquals('valid', $check['status']);
+
+        $process->stop();
+
+        /*
+         * Request certificate
+         */
+        $csr = new CertificateRequest(new DistinguishedName('acmephp.com'), (new KeyPairGenerator())->generateKeyPair());
+        $response = $this->client->requestCertificate('acmephp.com', $csr);
+
+        $this->assertInstanceOf('AcmePhp\\Ssl\\CertificateResponse', $response);
```

Comment:
IMO `CertificateResponse::class` is much more readable and easy to use thanks to IDE